patches/net/minecraft/world/entity/LivingEntity.java.patch

```diff
@@ -350,23 +406,43 @@
+ 
+                 return p_21194_;
+@@ -1699,11 +_,14 @@
  
      protected void actuallyHurt(DamageSource p_21240_, float p_21241_) {
          if (!this.isInvulnerableTo(p_21240_)) {
-+            p_21241_ = net.neoforged.neoforge.common.CommonHooks.onLivingHurt(this, p_21240_, p_21241_);
-+            if (p_21241_ <= 0) return;
-             p_21241_ = this.getDamageAfterArmorAbsorb(p_21240_, p_21241_);
-             p_21241_ = this.getDamageAfterMagicAbsorb(p_21240_, p_21241_);
-             float f1 = Math.max(p_21241_ - this.getAbsorptionAmount(), 0.0F);
-@@ -1708,6 +_,7 @@
+-            p_21241_ = this.getDamageAfterArmorAbsorb(p_21240_, p_21241_);
+-            p_21241_ = this.getDamageAfterMagicAbsorb(p_21240_, p_21241_);
+-            float f1 = Math.max(p_21241_ - this.getAbsorptionAmount(), 0.0F);
+-            this.setAbsorptionAmount(this.getAbsorptionAmount() - (p_21241_ - f1));
+-            float f = p_21241_ - f1;
++            this.damageContainer.getLast().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ARMOR, this.damageContainer.getLast().getNewDamage() - this.getDamageAfterArmorAbsorb(p_21240_, this.damageContainer.getLast().getNewDamage()));
++            this.getDamageAfterMagicAbsorb(p_21240_, this.damageContainer.getLast().getNewDamage());
++            this.damageContainer.getLast().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION, this.getAbsorptionAmount());
++            float f1 = Math.max(this.damageContainer.getLast().getNewDamage() - this.damageContainer.getLast().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION), 0.0F);
++            this.setAbsorptionAmount(this.damageContainer.getLast().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION) - (this.damageContainer.getLast().getNewDamage() - f1));
++            f1 = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainer.getLast());
++            if (f1 <= 0) return;
++            float f = this.damageContainer.getLast().getNewDamage() - f1;
+             if (f > 0.0F && f < 3.4028235E37F && p_21240_.getEntity() instanceof ServerPlayer serverplayer) {
                  serverplayer.awardStat(Stats.DAMAGE_DEALT_ABSORBED, Math.round(f * 10.0F));
              }
- 
-+            f1 = net.neoforged.neoforge.common.CommonHooks.onLivingDamage(this, p_21240_, f1);
-             if (f1 != 0.0F) {
-                 this.getCombatTracker().recordDamage(p_21240_, f1);
+@@ -1713,6 +_,8 @@
                  this.setHealth(this.getHealth() - f1);
+                 this.setAbsorptionAmount(this.getAbsorptionAmount() - f1);
+                 this.gameEvent(GameEvent.ENTITY_DAMAGE);
++                this.onDamageTaken(this.damageContainer.getLast());
++                net.neoforged.neoforge.common.CommonHooks.onLivingDamagePost(this, this.damageContainer.getLast());
```

Comment:
I know we already discussed this, but I just remembered why I specifically wanted the `Post` event __outside__ of these braces in the first place.

Because this handling is done inside of `actuallyHurt`, the remainder of the handling that happens in `hurt` does not take this value into account in any way, and proceeds based on the value that was initially passed to hurt (why, Mojang?)
This functionally means that all the damage modifications after the shield block __do not__ have an effect on the resultant triggering of knockback, hurt marking, etc.

TL;DR: The attack is still doing things after this, even if the value is 0. This event should be called after the closing brace of this section instead, so that callers can still hook successful attacks, even if the resultant "damage" is 0

I'm not sure if the same should apply to `onDamageTaken`, I'm open to debate on that since I can see valid justifications either way